[Draft] Testing json schema

Krupa2788 opened the pull request:
- Added TypeScript interface for data structure
- Automatic generation of JSON schemas from TypeScript types using commands
- Created validation.js utility file to reuse validation functions
- Added SCHEMA_VALIDATION documentation file
- **Runtime Validation**: AJV-powered validation of JSON data against schemas
- Added schema files and runtime validation for the components: Button, Footer
- **CLI Tools**: Command-line scripts for validation and testing

Krupa2788 committed:
- Added TypeScript interface for data structure
- Automatic generation of JSON schemas from TypeScript types using commands
- **Runtime Validation**: AJV-powered validation of JSON data against schemas
- **CLI Tools**: Command-line scripts for validation and testing

Krupa2788 committed:
Updated package-lock.json

Krupa2788 committed:
- Fixed git lint error by making code compatible with node 19
- Added footer.types.ts and footer.schema.js
- Created validation.js utility file to reuse validation functions
- Updated SCHEMA_VALIDATION documation file

Krupa2788 committed:
Merge branch 'develop' into testing-json-schema

longhssq commented on src/utils/validation.js:
```diff
@@ -0,0 +1,113 @@
+ * @returns {Object} - Validation result with isValid boolean and errors array
+ */
+export function validateData(data, schema, componentName = "Component") {
+  const validate = ajv.compile(schema);
```
We should use a cache approach for multiple components' schema compiling. The execution here should only be called once in our app
```
let validate = ajv.getSchema(componentName)
if (!validate) {
  ajv.addSchema(schema, componentName)
  validate = ajv.getSchema(componentName)
}
```
FYI: https://ajv.js.org/guide/managing-schemas.html#using-ajv-instance-cache

longhssq commented on scripts/generate-schemas.js:
```diff
@@ -0,0 +1,87 @@
+    const command = `npx ts-json-schema-generator --path ${inputFile} --type ${type} --tsconfig tsconfig.schemas.json --out ${outputFile}`;
+
+    console.log(`Generating schema for ${type}...`);
+    execSync(command, {
```
AI is right, `execFileSync` seems better for sync blocking execution

Krupa2788 committed:
refactor: reorganize JS utilities to src/js folder

- Move validation.js and generate-schemas.js to src/js/
- Update all import paths and script references
- Update documentation to reflect new structure

stvp-qld commented:
Closing and moving work to backlog - future implementation
